File: desktop/mathmark/reminders.py
"""
Напоминания, навешенные на файл.

Хранятся отдельным текстовым файлом рядом с настройками — внутрь `.md`
не пишется ничего:

    Линал/собственные.md|daily|19:00|повторить линал
    Матан.md|weekly|1|09:30|разобрать Тейлора
    Шпора.md|once|2026-08-05T19:00|перед экзаменом

Строку можно поправить руками. Раз файл лежит не в папке с математикой,
синхронизация его не переносит: на каждом устройстве свои напоминания.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, datetime, time, timedelta
from pathlib import Path

DAILY, WEEKLY, ONCE = "daily", "weekly", "once"


@dataclass(frozen=True)
class Reminder:
    path: str          # путь файла относительно папки с математикой
    repeat: str        # daily | weekly | once
    at: time           # время суток
    text: str
    weekday: int = 0   # для weekly: 1 — понедельник … 7 — воскресенье
    on: date | None = None   # для once

    def line(self) -> str:
        if self.repeat == DAILY:
            return f"{self.path}|daily|{self.at:%H:%M}|{self.text}"
        if self.repeat == WEEKLY:
            return f"{self.path}|weekly|{self.weekday}|{self.at:%H:%M}|{self.text}"
        when = datetime.combine(self.on or date.today(), self.at)
        return f"{self.path}|once|{when.isoformat(timespec='minutes')}|{self.text}"

    def due_on(self, day: date) -> bool:
        """Сработает ли в этот день."""
        if self.repeat == DAILY:
            return True
        if self.repeat == WEEKLY:
            return day.isoweekday() == self.weekday
        return self.on == day

    def next_after(self, moment: datetime) -> datetime | None:
        """Ближайшее срабатывание строго после указанного момента."""
        if self.repeat == ONCE:
            when = datetime.combine(self.on, self.at) if self.on else None
            return when if when and when > moment else None
        day = moment.date()
        for step in range(0, 8):
            d = day + timedelta(days=step)
            if not self.due_on(d):
                continue
            when = datetime.combine(d, self.at)
            if when > moment:
                return when
        return None


def parse(text: str) -> list[Reminder]:
    out: list[Reminder] = []
    for raw in text.split("\n"):
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        parts = line.split("|")
        try:
            if len(parts) == 4 and parts[1] == DAILY:
                out.append(Reminder(parts[0], DAILY, _time(parts[2]), parts[3]))
            elif len(parts) == 5 and parts[1] == WEEKLY:
                wd = int(parts[2])
                if not 1 <= wd <= 7:
                    continue
                out.append(Reminder(parts[0], WEEKLY, _time(parts[3]), parts[4], weekday=wd))
            elif len(parts) == 4 and parts[1] == ONCE:
                when = datetime.fromisoformat(parts[2])
                out.append(Reminder(parts[0], ONCE, when.time(), parts[3], on=when.date()))
        except (ValueError, IndexError):
            continue
    return out


def _time(s: str) -> time:
    h, m = s.split(":")
    return time(int(h), int(m))


def dump(items: list[Reminder]) -> str:
    return "".join(r.line() + "\n" for r in items)


def load(file: Path) -> list[Reminder]:
    try:
        return parse(file.read_text(encoding="utf-8"))
    except OSError:
        return []


def save(file: Path, items: list[Reminder]) -> None:
    try:
        file.parent.mkdir(parents=True, exist_ok=True)
        file.write_text(dump(items), encoding="utf-8")
    except OSError:
        pass


def for_day(items: list[Reminder], day: date) -> list[Reminder]:
    """Что сработает в этот день — по возрастанию времени."""
    return sorted((r for r in items if r.due_on(day)), key=lambda r: r.at)

File: android/app/src/main/java/dev/yury/mathmark/Reminders.kt
package dev.yury.mathmark

import java.io.File
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

/**
 * Напоминания, навешенные на файл.
 *
 * Хранятся отдельным текстовым файлом рядом с настройками — внутрь `.md`
 * не пишется ничего:
 *
 *     Линал/собственные.md|daily|19:00|повторить линал
 *     Матан.md|weekly|1|09:30|разобрать Тейлора
 *     Шпора.md|once|2026-08-05T19:00|перед экзаменом
 *
 * Раз файл лежит не в папке с математикой, синхронизация его не переносит:
 * на каждом устройстве свои напоминания.
 *
 * Точный повтор `reminders.py` настольной версии.
 */

enum class Repeat { DAILY, WEEKLY, ONCE }

data class Reminder(
    val path: String,
    val repeat: Repeat,
    val at: LocalTime,
    val text: String,
    val weekday: Int = 0,          // 1 — понедельник … 7 — воскресенье
    val on: LocalDate? = null,     // для однократного
) {
    fun line(): String = when (repeat) {
        Repeat.DAILY -> "$path|daily|${hhmm()}|$text"
        Repeat.WEEKLY -> "$path|weekly|$weekday|${hhmm()}|$text"
        Repeat.ONCE -> "$path|once|${LocalDateTime.of(on ?: LocalDate.now(), at)}|$text"
    }

    private fun hhmm(): String = "%02d:%02d".format(at.hour, at.minute)

    fun dueOn(day: LocalDate): Boolean = when (repeat) {
        Repeat.DAILY -> true
        Repeat.WEEKLY -> day.dayOfWeek.value == weekday
        Repeat.ONCE -> on == day
    }

    /** Ближайшее срабатывание строго после указанного момента. */
    fun nextAfter(moment: LocalDateTime): LocalDateTime? {
        if (repeat == Repeat.ONCE) {
            val when0 = on?.let { LocalDateTime.of(it, at) } ?: return null
            return if (when0.isAfter(moment)) when0 else null
        }
        for (step in 0..7) {
            val d = moment.toLocalDate().plusDays(step.toLong())
            if (!dueOn(d)) continue
            val w = LocalDateTime.of(d, at)
            if (w.isAfter(moment)) return w
        }
        return null
    }
}

object Reminders {

    fun parse(text: String): List<Reminder> {
        val out = ArrayList<Reminder>()
        for (raw in text.split("\n")) {
            val line = raw.trim()
            if (line.isEmpty() || line.startsWith("#")) continue
            val p = line.split("|")
            runCatching {
                when {
                    p.size == 4 && p[1] == "daily" ->
                        out.add(Reminder(p[0], Repeat.DAILY, time(p[2]), p[3]))
                    p.size == 5 && p[1] == "weekly" -> {
                        val wd = p[2].toInt()
                        if (wd in 1..7) out.add(Reminder(p[0], Repeat.WEEKLY, time(p[3]), p[4], weekday = wd))
                    }
                    p.size == 4 && p[1] == "once" -> {
                        val w = LocalDateTime.parse(p[2])
                        out.add(Reminder(p[0], Repeat.ONCE, w.toLocalTime(), p[3], on = w.toLocalDate()))
                    }
                }
            }
        }
        return out
    }

    private fun time(s: String): LocalTime {
        val (h, m) = s.split(":")
        return LocalTime.of(h.toInt(), m.toInt())
    }

    fun dump(items: List<Reminder>): String = items.joinToString("") { it.line() + "\n" }

    fun load(file: File): List<Reminder> =
        runCatching { parse(file.readText(Charsets.UTF_8)) }.getOrDefault(emptyList())

    fun save(file: File, items: List<Reminder>) {
        runCatching {
            file.parentFile?.mkdirs()
            file.writeText(dump(items), Charsets.UTF_8)
        }
    }

    /** Что сработает в этот день — по возрастанию времени. */
    fun forDay(items: List<Reminder>, day: LocalDate): List<Reminder> =
        items.filter { it.dueOn(day) }.sortedBy { it.at }

    /** Понятное человеку описание: «каждый день в 19:00». */
    fun describe(r: Reminder): String {
        val time = "%02d:%02d".format(r.at.hour, r.at.minute)
        return when (r.repeat) {
            Repeat.DAILY -> L.f("rem.everyDay", time)
            Repeat.WEEKLY -> L.f("rem.everyWeek", L["rem.day" + r.weekday], time)
            Repeat.ONCE -> L.f("rem.once", r.on?.toString().orEmpty(), time)
        }
    }
}
